modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala

```diff
@@ -1467,14 +1509,32 @@ private[smithy4s] class SchemaVisitorJCodec(
         { (meta: scala.collection.Map[String, Any]) =>
           meta.foreach(kv => buffer.put(kv._1, kv._2))
           val stage2 = new VectorBuilder[Any]
+          val unknownValue =
+            if (keepUnknown && unknownValues.nonEmpty) {
+              val builder = Map.newBuilder[String, Document]
+              builder ++= unknownValues
+              builder.result()
+            } else null
+
           fields.foreach { case (f, jsonLabel, default) =>
             stage2 += {
-              val value = buffer.get(f.label)
-              if (value == null) {
-                if (default == null)
-                  cursor.requiredFieldError(jsonLabel, jsonLabel)
-                else default
-              } else value
+              if (isLabelForUnknownFields(jsonLabel)) {
+                if (unknownValue == null) {
+                  if (default == null) {
+                    if (f.isRequired) Map.empty
+                    else None
+                  } else default
+                } else {
+                  if (f.isRequired) unknownValue else Some(unknownValue)
```

Comment:
you cannot make the assumption that `!required` implies `scala.Option` : we're planning on relaxing our Schema model to allow for other ways of modelling optionality (to cater for `java.util.Optional`, for instance).